### ISSUE TITLE
Any idea why I lose my weights when I have my tensors in a list?

### ISSUE THREAD

**Issue by asieradzk** [dotnet/TorchSharp#1100]
Perhaps it will be better to show what I'm trying to do with an example.
So please have a look:

This is how normally I would make a class to instantiate a neural network:

```cs
  public sealed class DQN1D : DQNNET
    {
        private readonly Module<Tensor, Tensor> fc1, fc2, fc3;
        private readonly Module<Tensor, Tensor> head;

        /// <summary>
        /// Initializes a new instance of the DQN1D class.
        /// </summary>
        /// <param name="name">The name of the network.</param>
        /// <param name="obsSize">The size of the observation space, i.e., the number of inputs to the network.</param>
        /// <param name="width">The number of neurons in each hidden layer.</param>
        /// <param name="actionSize">The size of the action space, i.e., the number of outputs from the network.</param>
        public DQN1D(string name, int obsSize, int width, int actionSize) : base(name)
        {
            fc1 = Linear(obsSize, width);
            fc2 = Linear(width, width);
            fc3 = Linear(width, width);
            head = Linear(width, actionSize);

            RegisterComponents();
        }

        public override Tensor forward(Tensor x)
        {
            if (x.dim() == 1)
            {
                x = x.unsqueeze(0);
            }

            x = functional.relu(fc1.forward(x));
            x = functional.relu(fc2.forward(x));
            x = functional.relu(fc3.forward(x));
            return head.forward(x);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                fc1.Dispose();
                fc2.Dispose();
                fc3.Dispose();
                head.Dispose();
                ClearModules();
            }

            base.Dispose(disposing);
        }
    }
    ```
    
    This works nicely but I wans wondering how I can abstract away the depth (or number of layers) somewhere up my pipeline, so I have something like this:
    
    ```cs
    public sealed class DQN1D : DQNNET
    {
        private readonly List<Module<Tensor, Tensor>> fcs;
        private readonly Module<Tensor, Tensor> head;

        /// <summary>
        /// Initializes a new instance of the DQN1D class.
        /// </summary>
        /// <param name="name">The name of the network.</param>
        /// <param name="obsSize">The size of the observation space, i.e., the number of inputs to the network.</param>
        /// <param name="width">The number of neurons in each hidden layer.</param>
        /// <param name="actionSize">The size of the action space, i.e., the number of outputs from the network.</param>
        /// <param name="depth">The number of fully connected layers.</param>
        public DQN1D(DeviceType myDevice, string name, int obsSize, int width, int actionSize, int depth=1) : base(name)
        {
            fcs = new List<Module<Tensor, Tensor>>();

            // First layer
            fcs.Add(Linear(obsSize, width).to(myDevice));

            // Middle layers
            for (int i = 1; i < depth - 1; i++)
            {
                fcs.Add(Linear(width, width).to(myDevice));
            }

            // Head layer
            head = Linear(width, actionSize).to(myDevice);

            RegisterComponents();
        }

        public override Tensor forward(Tensor x)
        {
            if (x.dim() == 1)
            {
                x = x.unsqueeze(0);
            }

            foreach (var fc in fcs)
            {
                x = functional.relu(fc.forward(x));
            }

            return head.forward(x);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                foreach (var fc in fcs)
                {
                    fc.Dispose();
                }
                head.Dispose();
                ClearModules();
            }

            base.Dispose(disposing);
        }
    }

```

Problem is, this is completely not the same as the example above and during training I keep losing my weights, can't really see what I'm doing wrong here. A second pair of eyes, please?

**Comment by asieradzk** [comment]
> There's a `ModuleList` in torch, which the framework is aware of and can treat correctly. `RegisterComponents` will see a field with a regular list and say "that's not a module..." but it will do the right thing with `ModuleList` There's also a `ModuleDict`.
> 
> The reason this affects training is that the optimizer will ask for the parameters of the module, and the parameters (weights) of the modules in the list will be skipped, so the optimizer will not use them when doing its job. The backprop should still be computing the gradients, but they won't be used during optimization.

Thank you for the answer. The ModuleList indeed is what I was looking for.

**Comment by asieradzk** [comment]
> A couple of other observations that this second pair of eyes spied:
> 
> 1. You really should use a dispose scope inside your `forward` implementation. That's important given the number of temporaries you're creating. See 'Technique 3' in https://github.com/dotnet/TorchSharp/wiki/Memory-Management
> 2. ModuleList should work nicely, but for this particular situation, adding the dynamic `Linear` layers to a Sequence, instead, will result in _slightly_ better memory performance. There's a `ReLU` block that you can insert after each `Linear` instance, too. The only thing you need to keep outside the sequence is the call to `unsqueeze`. The Sequence implementation releases memory as soon as possible, whereas using a dispose scope (which you really should be doing if you choose not to use Sequence, see above :-)) will wait until the end of the `forward` to do so. Plus, you don't have to have the loop, you just call the Sequence as a module.

**Comment by asieradzk** [comment]
> A couple of other observations that this second pair of eyes spied:
> 
> 1. You really should use a dispose scope inside your `forward` implementation. That's important given the number of temporaries you're creating. See 'Technique 3' in https://github.com/dotnet/TorchSharp/wiki/Memory-Management
> 2. ModuleList should work nicely, but for this particular situation, adding the dynamic `Linear` layers to a Sequence, instead, will result in _slightly_ better memory performance. There's a `ReLU` block that you can insert after each `Linear` instance, too. The only thing you need to keep outside the sequence is the call to `unsqueeze`. The Sequence implementation releases memory as soon as possible, whereas using a dispose scope (which you really should be doing if you choose not to use Sequence, see above :-)) will wait until the end of the `forward` to do so. Plus, you don't have to have the loop, you just call the Sequence as a module.

Thank you for advice. idd I do not adehere to the best memory management practice, perhaps I didn't notice since working on a powerhouse machine.

I am not sure how to properly use dispose scope inside my forward method. Could you give me a pseudo-code example before I close the issue? 🥺

**Comment by NiklasGustafsson** [comment]
Like so:

```C#
        public override Tensor forward(Tensor x)
        {
            using var _ = torch.NewDisposeScope();
            if (x.dim() == 1)
            {
                x = x.unsqueeze(0);
            }

            foreach (var fc in fcs)
            {
                x = functional.relu(fc.forward(x));
            }

            return head.forward(x).MoveToOuterDisposeScope();
        }
```

**Comment by asieradzk** [comment]
> Like so:
> 
> ```cs
>         public override Tensor forward(Tensor x)
>         {
>             using var _ = torch.NewDisposeScope();
>             if (x.dim() == 1)
>             {
>                 x = x.unsqueeze(0);
>             }
> 
>             foreach (var fc in fcs)
>             {
>                 x = functional.relu(fc.forward(x));
>             }
> 
>             return head.forward(x).MoveToOuterDisposeScope();
>         }
> ```

Now I see it! Thanks
